### PR TITLE
Boat Module

### PR DESCRIPTION
**Pull request by PerFrivik**
### Solved Problem
This PR integrates the Boat module with a new architecture that supports modular components such as BoatControl, BoatGuidance, and BoatKinematics. This architecture facilitates the integration with different drive systems like Differential Drive and Ackermann Drive. The module is designed for scalability, allowing for future adaptations to various types of boats by adjusting components like kinematics to suit specific needs.

### Solution
- Added support for new architecture in Boat module.
- Refactored existing code to improve modularity.

### Changelog Entry
For release notes:
```
Feature/Bugfix XYZ
New parameter: XYZ_Z
Documentation: Need to clarify page ... / done, read docs.px4.io/...
```

### Alternatives
We could also ...

### Test coverage
- Unit/integration test: ...
- Hardware test: https://review.px4.io/plot_app?log=b9f5f965-4b3f-4131-9ae3-92e2e4860261

### Context
Related links, screenshot before/after, video

### CODE REVIEW

**Comment by igalloway** [comment]
Hi @PerFrivik 
Would you be able to share a bit more about your setup? What type of boat are you testing on? I would be interested in attempting to replicate a boat config using  our NXP RT1176 based VMU/FMU. (NXP ref design similar to Holybro V6X-RT)
https://docs.px4.io/main/en/flight_controller/pixhawk6x-rt.html

**Comment by PerFrivik** [comment]
> Hi @PerFrivik Would you be able to share a bit more about your setup? What type of boat are you testing on? I would be interested in attempting to replicate/build up a boat config for testing using our NXP RT1176 based VMU/FMU. (NXP ref design similar to Holybro V6X-RT) https://docs.px4.io/main/en/flight_controller/pixhawk6x-rt.html

Yes of course! I used the setup that @junwoo091400 setup quite a while ago, which can be found [here](https://discuss.px4.io/t/rc-speed-boat-with-px4-episode-1-using-px4-to-control-the-boat/28429).

**Comment by PerFrivik** [comment]
Keep in mind this PR is subject to change, I will most likely replace the L1 with a pure pursuit.

**Comment by hamishwillee** [comment]
@PerFrivik We'll need to discuss then when it goes in for docs.

**Comment by PerFrivik** [comment]
> @PerFrivik We'll need to discuss then when it goes in for docs.

Yes I agree, I'll tag you then!

File: msg/BoatSetpoint.msg [hunk 1]
```diff
@@ -0,0 +1,8 @@
+uint64 timestamp # time since system start (microseconds)
+
+float32 speed # [m/s] collective roll-off speed in body x-axis
+bool closed_loop_speed_control # true if speed is controlled using estimator feedback, false if direct feed-forward
+float32 yaw_rate # [rad] yaw_rate
+bool closed_loop_yaw_rate_control # true if rudder angle is controlled using encoder feedback, false if direct feed-forward
+
+# TOPICS boat_setpoint boat_control_output
```
[the remark this answers is not in the document]
Comment (PerFrivik) [reply msg/BoatSetpoint.msg:8]: Sounds good, thank you for your input! I'm doing quite the revamp on the controller and this was exactly one of my questions.

**Pull request closed without merging**